libsolidity/ast/AST.h

```diff
@@ -540,6 +540,27 @@ class ContractDefinition: public Declaration, public StructurallyDocumented
 	util::LazyInit<std::vector<EventDefinition const*>> m_interfaceEvents;
 };
 
+/**
+ * A sequence of identifiers separated by dots
+ */
+class IdentifierPath: public ASTNode
+{
+public:
+	IdentifierPath(int64_t _id, SourceLocation const& _location, std::vector<ASTString> _path):
+		ASTNode(_id, _location), m_path(std::move(_path)) {}
+
+	std::vector<ASTString> const& path() const { return m_path; }
+	IdentifierPathAnnotation& annotation() const override
+	{
+		return initAnnotation<IdentifierPathAnnotation>();
+	}
+
+	void accept(ASTVisitor& _visitor) override;
+	void accept(ASTConstVisitor& _visitor) const override;
+private:
+	std::vector<ASTString> m_path;
+};
+
 class InheritanceSpecifier: public ASTNode
```

Comment:
Since we are doing this as a breaking change, I think we should do it properly: Everything that is not really a UserDefinedTypeName should not be a UserDefinedTypeName. In the InheritanceSpecifier, for example, we are mainly referencing a contract, but it is not used a the type of a variable in this case, so it should be IdentifierPath.

Reply:
I just searched AST.h and I think that every explicit mention of UserDefinedTypeName should be IdentifierPath instead. UserDefinedTypeName should only be used implicitly through its base class TypeName.

Reply:
> we are mainly referencing a contract, but it is not used a the type of a variable in this case, so it should be IdentifierPath.

I am not really following this. A contract is a user defined typename and an Inheritance specifier can only have those, why should it be something more generic in this case?

When I went through the AST I tried to think which places can have things other than types and found that only the modifier invocation with the super constructor call really matches that..